### ISSUE TITLE
How to send producer timestamp

### ISSUE THREAD

**Issue by ynaka81** [awslabs/amazon-kinesis-video-streams-producer-sdk-cpp#35]
When I run `kinesis_video_gstreamer_sample_app`, I can see video preview in server time stamps mode on AWS Web console. But in producer time stamps mode, I can't see video preview. I also checked producer time stamp with `amazon-kinesis-video-streams-parser-library/src/main/java/com/amazonaws/kinesisvideo/parser/utilities/FragmentMetadata.java:FragmentMetadata.getProducerSideTimetampAsDate()`. It returned `Thu Feb 12 01:00:24 JST 1970` which seems to be incorrect. I thought it is caused by `steady_clock` and change it to `system_clock` but the result was still incorrect.
```diff
diff --git a/kinesis-video-pic/src/client/src/StreamState.cpp b/kinesis-video-pic/src/client/src/StreamState.cpp
index cbc684b..3537767 100644
--- a/kinesis-video-pic/src/client/src/StreamState.cpp
+++ b/kinesis-video-pic/src/client/src/StreamState.cpp
@@ -2,6 +2,7 @@
 * Implementation of a stream states callbacks
 */

+#include <chrono>
#define LOG_CLASS "StreamState"
#include "Include_i.h"

@@ -604,7 +605,7 @@ STATUS executePutStreamState(UINT64 customData, UINT64 time)
            pKinesisVideoClient->clientCallbacks.customData,
            pKinesisVideoStream->streamInfo.name,
            (PCHAR) MKV_CONTAINER_TYPE_STRING,
-            pKinesisVideoClient->clientCallbacks.getCurrentTimeFn(pKinesisVideoClient->clientCallbacks.customData),
+            std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::system_clock::now().time_since_epoch()).count() / DEFAULT_TIME_UNIT_IN_NANOS,
            pKinesisVideoStream->streamInfo.streamCaps.absoluteFragmentTimes,
            pKinesisVideoStream->streamInfo.streamCaps.fragmentAcks,
            pKinesisVideoStream->streamingEndpoint,
```

So how can I send producer timestamp to kinesis video stream?

**Comment by zhiyua-git** [comment]
Hi ynaka81,

By default, the producer timestamp is provided by you. Can you check what timestamp you put [here](https://github.com/awslabs/amazon-kinesis-video-streams-producer-sdk-cpp/blob/2aef898396672e7ccc7dcdd13f8c56e362e4d331/kinesis-video-gst-demo/kinesis_video_gstreamer_sample_rtsp_app.cpp#L163)? Timestamp is from the device itself, maybe the device clock is set to some old time.

Please let me know how it goes. Thanks.
-ZH

**Comment by ynaka81** [comment]
Hi zhiyua-git. Thank you.
I changed the source code to print `pts` and `dts` as following.
```diff
diff --git a/kinesis-video-gst-demo/kinesis_video_gstreamer_sample_app.cpp b/kinesis-video-gst-demo/kinesis_video_gstreamer_sample_app.cpp
index 0f3c29a..478e9c4 100644
--- a/kinesis-video-gst-demo/kinesis_video_gstreamer_sample_app.cpp
+++ b/kinesis-video-gst-demo/kinesis_video_gstreamer_sample_app.cpp
@@ -6,6 +6,7 @@
 #include "KinesisVideoProducer.h"
 #include <vector>
 #include <stdlib.h>
+#include <iostream>
 
 using namespace std;
 using namespace com::amazonaws::kinesis::video;
@@ -157,6 +158,7 @@ void create_kinesis_video_frame(Frame *frame, const nanoseconds &pts, const nano
 
 bool put_frame(shared_ptr<KinesisVideoStream> kinesis_video_stream, void *data, size_t len, const nanoseconds &pts, const nanoseconds &dts, FRAME_FLAGS flags) {
     Frame frame;
+    cout << "pts:" << pts.count() << ", dts:" << dts.count() << endl;
     create_kinesis_video_frame(&frame, pts, dts, flags, data, len);
     return kinesis_video_stream->putFrame(frame);
 }
```
I got `pts:3600002555339025, dts:3600002555339025`. It seems to be incorrect. Even if I use `system_clock` instead of `steady_clock`, the result was not changed.
I found that the timestamp from gstreamer is wrong. It might be related with I used web-cam.

**Comment by zhiyua-git** [comment]
Hi yanka81,

Would you mind providing more information for letting us solve your issue?

* What USB webcam are you using? Can you provide us the model and manufacturer of your USB webcam?to run the gstream app? Is it getting h264 frames directly or from h264 ecoder in gstreamer? Just wanted to check if the GStreamer getting the h264 encoded frames from the camera or not.
* What kind of data consumption are you planning to do on the data, playback or video analysis?

Thanks,
-ZH

**Comment by ynaka81** [comment]
Hi zhiyua-git,
- I used equipped camera of my ThinkPad T460s (https://www3.lenovo.com/us/en/laptops/thinkpad/thinkpad-t-series/ThinkPad-T460s/p/22TP2TT460S).
- I just ran `$ ./kinesis_video_gstreamer_sample_app example` to send video stream to KVS.
- I'm planning to use it for video analysis. As a first step, I checked video preview on AWS Web console.
- I can see video preview in server time stamps mode, but can't see it in producer time stamps mode.

Let me know if you need more information. Thanks.

**Comment by zhiyua-git** [comment]
Hi ynaka81, 

Can you also provide some details about your stream? Like stream name, account Id, and request Id in log?

Request id will look like "DEBUG / KinesisVideo: Received ACK bits: x-amzn-RequestId: 8b9ea9ef-12bd-11e8-8d8a-15e16ecsample" in the log.

It will be great if you can send this information to our support email: kinesis-video-support@amazon.com.

Thanks,
-ZH

**Comment by ynaka81** [comment]
Hi zhiyua-git,
Thank you. I sent to your support email.

**Comment by vijaygj** [comment]
Hello ynaka81@,

Can you please send us more details about following failure on the same email address?

> I can see video preview in server time stamps mode, but can't see it in producer time stamps mode.

We see that the producer time stamps are being assigned to the fragments under your stream but may be the time you are using to "search" is not in sync with time assigned to each fragment. Can you please send the screen shot of the playback issue? We can check further. 

Thanks,
Vijay.

**Comment by ynaka81** [comment]
Hello vijaygj,
I sent to your support mail.

As you say, I think that the time is not in sync with time assigned to each fragment because the timestamp from gstreamer is already incorrect (cf. https://github.com/awslabs/amazon-kinesis-video-streams-producer-sdk-cpp/issues/35#issuecomment-366821607).

Oh, I might misunderstand. The producer time stamp is not the time like UTC or JST, but the uptime of producer, isn't it?

**Comment by vijaygj** [comment]
Thank you for the email. The producer time stamp is indeed local time on the camera device hence in millis since epoch and not uptime of the producer. We will check why producer SDK is not able to capture the correct local time and update the thread.

**Comment by ynaka81** [comment]
Hello vijaygj,
Thank you for your cooperation. I'm waiting for your update.

**Comment by zhiyua-git** [comment]
Hi ynaka81,

We just posted a fix in [release 1.2.1](https://github.com/awslabs/amazon-kinesis-video-streams-producer-sdk-cpp/releases/tag/1.2.1) for system_clock used by SDK related to producer timestamp playback in console. 

This should help. Can you give it a try?

Thanks,
-ZH

**Comment by ynaka81** [comment]
Hi zhiyua-git,

Thank you for fixing it. It works well!

**Comment by zhiyua-git** [comment]
Great to hear it works!

**Comment by wkentaro** [comment]
Hi, I encountered the same (or similar) problem, so could you give me some information to fix it?
I am using v1.3.1 of this SDK, and I can see the frame on web console with both producer and server timestamp option.
But producer timestamp differs from the actual time from 5 seconds to 10 minutes.
(the server timestamp also has ~5seconds time lag but I suppose it is the time spent for the upload)

Note that I have checked the times are synchronized for both the clock in the video and the PC to which the USB camera is connected.

**server** timestamp (looks ok)

<img src=https://user-images.githubusercontent.com/4310419/39197306-24832e06-481f-11e8-88f7-08b52c2d52ea.png width=50% />

**producer** timestamp

time lag: ~10minutes
<img src=https://user-images.githubusercontent.com/4310419/39197376-44ee5972-481f-11e8-96ae-333a3e93e716.png width=50% />

time lag: ~5seconds
<img src="https://user-images.githubusercontent.com/4310419/39197513-a3c60882-481f-11e8-8e22-68eb484555f8.png" width="50%" />

**Comment by MushMal** [comment]
@wkentaro the console playback is a "live" playback and it's not "realtime". It is based on fragments being "assembled", persisted, indexed on the backend, then, downloaded to the console player where the browsers MSE is used to playback the MPEG fragments in the right order. Generally, the observed latency in the console playback is expected to be the fragment duration + ~3 seconds for the end-to-end. Could you please check what's your fragment duration?

As for the producer timestamp playback - it's not yet immediately clear as to what's going on. Is the playback starting 10 minutes after the start or it starts and then drifts? One thing that needs to be noted is that the browser will stop/pause the playback if the screen becomes inactive and will re-start from the paused time when the screen becomes active again.

Could you please provide more information? If this is a 10 minute delay to start the playback then we will need to dive deep and root cause this issue.

**Comment by wkentaro** [comment]
>  the console playback is a "live" playback and it's not "realtime"..

Yeah, I think I understand. It takes 5-10 seconds before the playback starts, and usually I observe `live-7seconds` at the right bottom in the console.

----

<img src=https://user-images.githubusercontent.com/4310419/39207949-532a32f0-483c-11e8-854c-9f05d9c89df0.png width=50% />

I expect the time in the left of this figure (the figure below, 0:28:32) and the time in the frame (0:28:27) would be same with the producer timestamp playback.

<img src=https://user-images.githubusercontent.com/4310419/39208022-9e2113dc-483c-11e8-8a1c-d4fd08787d97.png width=50% />

I also expect the time of the right bottom in the console and the frame in the console is synchronized. 
Maybe, my expectations are not right?

----
I have tried uploading video many times, and it was very rare case that I had 10 minutes delay, so maybe it could be caused by my mistake or browser cache or reload.
But I always have >5seconds delay I described above, and that's a critical issue for me.

**Comment by zhiyua-git** [comment]
@wkentaro Can you check the **Fragment duration** in the console page? The delay you will see in console would be your fragment duration + console player buffering time(~3 seconds). So 5-second delay in console you see would be quite common if your fragment duration is default 2-second setting.

If you want realtime playback scenario (), you will need some type of realtime player implementation.  You can refer to our realtime playback sample [KinesisVideoRendererExample](https://github.com/aws/amazon-kinesis-video-streams-parser-library#example) in our parser library.

Let me know if this helps. Thanks.

**Comment by wkentaro** [comment]
Ah, sorry. I overlooked the advice of @MushMal about Fragment duration.

I checked the Fragment duration on the console page as below.
It is 3.0 seconds, so it seems that it is working correctly as you said.
I will try the realtime player implementation. Thank you for your help!

<img width="1285" alt="screen shot 2018-04-25 at 17 59 06" src="https://user-images.githubusercontent.com/4310419/39235964-60ed90f0-48b2-11e8-9f35-2b0a7ab5e5e3.png">
